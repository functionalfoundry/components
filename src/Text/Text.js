/* @flow */
import React from 'react'
import Theme from 'js-theme'
import {
  Fonts,
} from '@workflo/styles'
import applyLayout from '../View/modules/applyLayout'
import applyNativeMethods from '../View/modules/applyNativeMethods'
import createDOMElement from '../View/modules/createDOMElement'

type AccessibilityRoleT = 'heading' | 'link'

type PropsT = {
  children?: React.Children,
  size: string,
  theme: Object,
  accessibilityRole: AccessibilityRoleT,
  onLayout: Function,
  onPress: Function,
  selectable: bool,
  singleLine: bool,
  element: string,
}

const defaultProps = {
  accessible: true,
  selectable: true,
  size: 'base',
  onPress: () => {},
  theme: {},
  element: 'span',
}

const Text = ({
  element,
  singleLine, // eslint-disable-line
  onLayout, // eslint-disable-line
  onPress,
  selectable, // eslint-disable-line
  theme,
  ...props
}: PropsT) => createDOMElement(element, {
  ...props,
  ...theme.text,
  onClick: onPress,
})

Text.defaultProps = defaultProps

const defaultTheme = ({
  singleLine,
  selectable,
  size,
}: PropsT) => ({
  text: {
    flex: '0 1 auto',
    color: 'inherit',
    display: 'inline',
    font: 'inherit',
    margin: 0,
    padding: 0,
    textDecorationLine: 'none',
    wordWrap: 'break-word',
    ...getFont(size),
    ...getSelectable(selectable),
    ...getLinesStyle(singleLine),
  },
})

const getFont = (size: string) => Fonts[size]

const getSelectable = (selectable: boolean) => {
  if (selectable) {
    return {}
  }
  return {
    userSelect: 'none',
  }
}

const getLinesStyle = (singleLine: bool) => {
  if (singleLine) {
    return {
      maxWidth: '100%',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
    }
  }
  return {}
}

export default Theme('Text', defaultTheme)(applyLayout(applyNativeMethods(Text)))
