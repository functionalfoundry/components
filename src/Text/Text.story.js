import React from 'react'
import { storiesOf } from '@kadira/storybook'
import Text from './Text'
import PreviewContainer from '../Preview/PreviewContainer'
import Preview from '../Preview'

storiesOf('Text', module)
  .add('Tiny', () => (
    <PreviewContainer
      shade='light'
    >
      <Preview
        label='Huge'
      >
        <Text
          size='huge'
        >
          {'Content Text'}
        </Text>
      </Preview>
      <Preview
        label='Large'
      >
        <Text
          size='large'
        >
          {'Content Text'}
        </Text>
      </Preview>
      <Preview
        label='Base'
      >
        <Text
          size='base'
        >
          {'Content Text'}
        </Text>
      </Preview>
      <Preview
        label='Small'
      >
        <Text
          size='small'
        >
          {'Content Text'}
        </Text>
      </Preview>
      <Preview
        label='Tiny'
      >
        <Text
          size='tiny'
        >
          {'Content Text'}
        </Text>
      </Preview>
    </PreviewContainer>
  ))
