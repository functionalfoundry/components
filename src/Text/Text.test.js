import React from 'react'
import renderer from 'react-test-renderer'
import Text from '.'

test('Renders with text', () => {
  const text = renderer.create(
    <Text>{'Hello'}</Text>
  ).toJSON()
  expect(text).toMatchSnapshot()
})
