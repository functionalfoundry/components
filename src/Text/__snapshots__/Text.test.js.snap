exports[`test Renders with text 1`] = `
<div
  className="view_1ps7m7i text_ropioz"
  data-test-id={undefined}
  size="base">
  Hello
</div>
`;
